src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/Handlers/TransportCreation/TcpTransportCreationHandler.cs

```diff
@@ -0,0 +1,198 @@
+            foreach (IPAddress ipAddress in ipAddresses)
+            {
+                ipEndpoint.Address = ipAddress;
+                try
+                {
+                    socket = await OpenSocket(ipEndpoint, isAsync, ct).ConfigureAwait(false);
+                    break;
+                }
+                catch(Exception e)
+                {
+                    socketOpenExceptions.Add(e);
+                }
+            }
+
+            // If no socket succeeded, throw
+            if (socket is null)
+            {
+                // If there are any socket exceptions in the collected exceptions, throw the first
+                // one. If there are not, collect all exceptions and throw them as an aggregate.
+                foreach (Exception exception in socketOpenExceptions)
+                {
+                    if (exception is SocketException)
+                    {
+                        throw exception;
+                    }
+                }
+
+                throw new AggregateException(socketOpenExceptions);
+            }
+
+            // Create the stream for the socket
+            return new NetworkStream(socket);
+        }
+
+        private async ValueTask<Socket> OpenSocket(IPEndPoint ipEndPoint, bool isAsync, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            var socket = new Socket(ipEndPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp) { Blocking = false };
+
+            // Enable keep-alive
+            socket.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveInterval, KeepAliveIntervalSeconds);
+            socket.SetSocketOption(SocketOptionLevel.Tcp, SocketOptionName.TcpKeepAliveTime, KeepAliveTimeSeconds);
+
+            try
+            {
+                if (isAsync)
+                {
+                    await socket.ConnectAsync(ipEndPoint, ct).ConfigureAwait(false);
+                }
+                else
+                {
+                    OpenSocketSync(socket, ipEndPoint, ct);
+                }
+            }
+            catch (Exception)
+            {
+                socket.Dispose();
+                throw;
+            }
+
+            // Connection is established
+            socket.Blocking = true;
+            socket.NoDelay = true;
+
+            return socket;
+        }
+
+        private void OpenSocketSync(Socket socket, IPEndPoint ipEndPoint, CancellationToken ct)
+        {
+            ct.ThrowIfCancellationRequested();
+
+            try
+            {
+                // Note: Although it seems logical to cancel if cancellation happens, we don't
+                //   need to do that here. Since the socket is set to be non-blocking, once we
+                //   call connect we'll throw and move onto polling where we continuously check
+                //   the cancellation token.
```

Comment:
I don't understand how it works. What does it mean "to cancel if cancellation happens"? What are possible scenarios when we arrive at the next line?

Reply:
Basically, if the cancellation token fires (ie, cancellation is requested), we would cancel connecting by disposing the socket. But since the socket is initialized as non-blocking, connect will immediately throw to indicate it would block (though connection still continues in the background). So there's no need to dispose the socket on cancellation until we move onto polling for a successful connection.

I've updated the comment to be a bit more clear about what's going on.